src/Adapter/Product/Create/CombinationCreator.php

```diff
@@ -0,0 +1,161 @@
+        $this->productRepository = $productRepository;
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param GroupedAttributeIds[] $groupedAttributeIdsList
+     *
+     * @return CombinationId[]
+     *@todo: multistore
+     */
+    public function createCombinations(ProductId $productId, array $groupedAttributeIdsList): array
+    {
+        $product = $this->productRepository->get($productId);
+        $generatedCombinations = $this->combinationGenerator->generate($this->formatScalarValues($groupedAttributeIdsList));
+
+        // avoid applying specificPrice on each combination.
+        SpecificPriceRule::disableAnyApplication();
+
+        $combinationIds = $this->addCombinations($product, $generatedCombinations);
+
+        Product::updateDefaultAttribute($product->id);
+        SpecificPriceRule::enableAnyApplication();
+        // apply all specific price rules at once after all the combinations are generated
+        SpecificPriceRule::applyAllRules([$product->id]);
+
+        return $combinationIds;
+    }
+
+    /**
+     * @param GroupedAttributeIds[] $groupedAttributeIdsList
+     *
+     * @return array<int, array<int, int>>
+     */
+    private function formatScalarValues(array $groupedAttributeIdsList): array
+    {
+        $groupedIdsList = [];
+        foreach ($groupedAttributeIdsList as $groupedAttributeIds) {
+            foreach ($groupedAttributeIds->getAttributeIds() as $attributeId) {
+                $groupedIdsList[$groupedAttributeIds->getAttributeGroupId()->getValue()][] = $attributeId->getValue();
+            }
+        }
+
+        return $groupedIdsList;
+    }
+
+    /**
+     * @param Product $product
+     * @param Traversable $generatedCombinations
+     *
+     * @return CombinationId[]
+     *
+     * @throws CombinationException
+     */
+    private function addCombinations(Product $product, Traversable $generatedCombinations): array
+    {
+        $product->setAvailableDate();
+
+        $addedCombinationIds = [];
+        foreach ($generatedCombinations as $generatedCombination) {
+            if ($product->productAttributeExists($generatedCombination, false, null, true)) {
+                continue;
+            }
+
+            $addedCombinationIds[] = $this->addCombination((int) $product->id, $generatedCombination);
+        }
+
+        return $addedCombinationIds;
+    }
+
+    /**
+     * @param int $productId
+     * @param int[] $generatedCombination
+     *
+     * @return CombinationId
+     */
+    private function addCombination(int $productId, array $generatedCombination): CombinationId
+    {
+        $newCombination = new Combination();
+        $newCombination->id_product = $productId;
+        $newCombination->default_on = 0;
+
+        $combinationId = $this->combinationRepository->add($newCombination);
+        $this->combinationRepository->saveProductAttributeAssociation($combinationId, $generatedCombination);
```

Comment:
As a safety, we could watch for `CannotAddCombinationException` when adding attributes to the combinations And if it is thrown the freshly created combination should be removed So that we don't end up with a half created Combination